src/python_testing/TC_DGGEN_2_4.py

```diff
@@ -0,0 +1,192 @@
+from chip.interaction_model import InteractionModelError
+from matter_testing_support import (MatterBaseTest, async_test_body, default_matter_test_main, matter_epoch_us_from_utc_datetime,
+                                    utc_datetime_from_matter_epoch_us, utc_datetime_from_posix_time_ms)
+from mobly import asserts
+
+logger = logging.getLogger(__name__)
+
+
+class TC_DGGEN_2_4(MatterBaseTest):
+    async def read_diags_attribute_expect_success(self, attribute):
+        cluster = Clusters.Objects.GeneralDiagnostics
+        return await self.read_single_attribute_check_success(endpoint=0, cluster=cluster, attribute=attribute)
+
+    async def read_timesync_attribute_expect_success(self, attribute):
+        cluster = Clusters.Objects.TimeSynchronization
+        return await self.read_single_attribute_check_success(endpoint=0, cluster=cluster, attribute=attribute)
+
+    async def set_time_in_timesync(self, current_time: int):
+        endpoint = 0
+        time_cluster = Clusters.Objects.TimeSynchronization
+        code = 0
+
+        try:
+            await self.send_single_cmd(cmd=time_cluster.Commands.SetUTCTime(UTCTime=current_time, granularity=time_cluster.Enums.GranularityEnum.kSecondsGranularity), endpoint=endpoint)
+        except InteractionModelError as e:
+            # The python layer discards the cluster specific portion of the status IB, so for now we just expect a generic FAILURE error
+            # see #26521
+            code = e.status
+
+        asserts.assert_true(code in [0, 1], "Unexpected error while trying to set the UTCTime")
+
+    async def send_time_snapshot_expect_success(self):
+        endpoint = 0
+        diags_cluster = Clusters.Objects.GeneralDiagnostics
+        code = 0
+
+        try:
+            response = await self.send_single_cmd(cmd=diags_cluster.Commands.TimeSnapshot(), endpoint=endpoint)
+        except InteractionModelError as e:
+            code = e.status
+
+        asserts.assert_equal(code, 0, "Expected success of TimeSnapshot.")
+        return response
+
+    @async_test_body
+    async def test_TC_GEN_2_4(self):
+        self.print_step(1, "Detect Time Synchronization UTCTime attribute presence")
+
+        root_descriptor = await self.default_controller.ReadAttribute(self.dut_node_id, [(0, Clusters.Descriptor)])
+        root_server_list = root_descriptor[0][Clusters.Descriptor][Clusters.Descriptor.Attributes.ServerList]
+        has_timesync = (Clusters.TimeSynchronization.id in root_server_list)
+
+        testvar_TimeSyncSupported = False
+
+        if has_timesync:
+            ts_attributes = await self.read_single_attribute(self.default_controller, node_id=self.dut_node_id, endpoint=0, attribute=Clusters.TimeSynchronization.Attributes.AttributeList)
+            has_utc_time = (Clusters.TimeSynchronization.Attributes.UTCTime.attribute_id in ts_attributes)
+
+            if has_utc_time:
+                testvar_TimeSyncSupported = True
```

Comment:
This matches the test plan, but I'm not sure it's correct there either. If you don't set the time before reading the UTC time attribute, then you're not guaranteed to have a non-null time. Then below, you won't go on to set the time. 

ie, I think step 3a should be moved up to before step 1, with a pre-check for the presence of the cluster.

Reply:
Will follow-up once testplan is fixed.